File: src/Web/wwwroot/js/ui-preferences.js
(() => {
    const themeModeCookie = "cv-theme-mode";
    const accentCookie = "cv-accent";
    const maxAge = 60 * 60 * 24 * 365;
    const themeModes = new Set(["light", "dark", "auto"]);
    const accents = new Set(["blue", "indigo", "teal", "rose", "amber"]);

    const getCookie = (name) => {
        const prefix = `${name}=`;
        const found = document.cookie.split("; ").find((part) => part.startsWith(prefix));
        return found ? decodeURIComponent(found.slice(prefix.length)) : null;
    };

    const setCookie = (name, value) => {
        document.cookie = `${name}=${encodeURIComponent(value)}; path=/; max-age=${maxAge}; samesite=lax`;
    };

    const resolveTheme = (mode) => {
        if (mode === "dark") {
            return "dark";
        }

        if (mode === "auto") {
            return window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light";
        }

        return "light";
    };

    const apply = (mode, accent) => {
        const themeMode = themeModes.has(mode) ? mode : "light";
        const color = accents.has(accent) ? accent : "blue";
        const root = document.documentElement;
        root.setAttribute("data-bs-theme", resolveTheme(themeMode));
        root.setAttribute("data-theme-mode", themeMode);
        root.setAttribute("data-accent", color);
    };

    const applyFromCookies = () => {
        apply(getCookie(themeModeCookie) || "light", getCookie(accentCookie) || "blue");
    };

    window.uiPreferences = {
        applyFromCookies,
        setThemeMode(mode) {
            const themeMode = themeModes.has(mode) ? mode : "light";
            setCookie(themeModeCookie, themeMode);
            apply(themeMode, getCookie(accentCookie) || "blue");
        },
        setAccent(accent) {
            const color = accents.has(accent) ? accent : "blue";
            setCookie(accentCookie, color);
            apply(getCookie(themeModeCookie) || "light", color);
        }
    };

    applyFromCookies();

    window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", () => {
        if ((getCookie(themeModeCookie) || "light") === "auto") {
            applyFromCookies();
        }
    });
})();

File: src/Web/Theming/UiPreferences.cs
namespace Web.Theming;

public static class UiPreferences
{
    public const string ThemeModeCookie = "cv-theme-mode";
    public const string AccentCookie = "cv-accent";
    public const string DefaultThemeMode = "light";
    public const string DefaultAccent = "blue";
    public const int CookieMaxAgeSeconds = 60 * 60 * 24 * 365;

    public static readonly string[] ThemeModes = ["light", "dark", "auto"];
    public static readonly string[] Accents = ["blue", "indigo", "teal", "rose", "amber"];

    public static string NormalizeThemeMode(string? value) =>
        ThemeModes.Contains(value, StringComparer.OrdinalIgnoreCase)
            ? value!.ToLowerInvariant()
            : DefaultThemeMode;

    public static string NormalizeAccent(string? value) =>
        Accents.Contains(value, StringComparer.OrdinalIgnoreCase)
            ? value!.ToLowerInvariant()
            : DefaultAccent;

    public static string ReadThemeMode(HttpContext? http) =>
        NormalizeThemeMode(http?.Request.Cookies[ThemeModeCookie]);

    public static string ReadAccent(HttpContext? http) =>
        NormalizeAccent(http?.Request.Cookies[AccentCookie]);

    public static string ResolveColorMode(string themeMode) =>
        themeMode == "dark" ? "dark" : "light";
}
